### PR TITLE
Add `load_meta` for faster metadata, improve DMI field handling, parsing optimizations

### PR DESCRIPTION
**Pull request by itsmeow**
## Breaking Changes

(minor) breaking API change, `Icon::load` now requires `Read + Seek` rather than just `Read` to be implemented on its reader. `BufReader` and `File` both implement `Seek`, though, so it's unlikely anything will have an issue with this.

### load_meta

`Icon::load` spends a lot of time calling `image::load_from_memory_with_format` and parsing IDAT chunks.

This PR adds `Icon::load_meta` and `RawDmi::load_meta`, which skip over any chunks that aren't the PNG header, IHDR, or zTXt. This makes it suitable for pure metadata (e.g. only reading the icon states, width, height, or DMI version) without spending any time doing image data parsing.

### Direct PNG Decoding

By converting from `image::load_from_memory_with_format` to a raw buffer decode with `PngDecoder`, the MTPC of `Icon::load` is reduced from 32.447ms to 21.175ms, making it significantly faster. The major savings are in `crop_imm` calls, which are now directly loaded from decoded PNG data by indexing, which is significantly faster.

### DMI Handling Improvements

`looping = 0` was found in at least one DMI I tested (I tested this on **every** BeeStation dmi file in the repository) so I made it correctly convert this to `Indefinitely`

It also appears a decent volume of DMIs did not have a width or height, so I made them optional fields and they default to 32x32 because that appears to be what BYOND does.

Fixes icon states that have names containing quotes or backslashes being parsed incorrectly. They will now be loaded and saved with escape characters handled automatically.

Also satisfies clippy by converting format strings to use templating. help.

### Performance

As you can see here almost all of `load`'s time is spent doing image parsing. We can just skip this!

image

Here's `load` after replacing the PNG decoder.

image

Here's `load_meta`

image

after replacing split_terminator with parse_dmi_line (I'm pretty sure it's almost *faster* despite handling quite a few more cases)

image

You can see that `load_meta` is ~75x faster (particularly for large icon files, smaller files it's only about 10x faster):

```
Icon::load_meta bench
---
Num calls: 68850
Total Call Duration (μs): 18896089
MTPC (μs): 274

Icon::load bench
---
Num calls: 1377
Total Call Duration (μs): 29158782
MTPC (μs): 21175
```

## Testing

I tested this on literally every image on TG and Bee, and it can load them just fine. I also added additional test images to cover edge cases (greyscale + alpha icons, extremely small icons that are well compressed)

### CODE REVIEW

File: tests/bench_dmi_load.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (ZeWaka) [inline tests/bench_dmi_load.rs]: commented out code

File: src/lib.rs [hunk 2]
```diff
@@ -6,10 +6,12 @@ pub mod icon;
 pub mod iend;
 pub mod ztxt;
 
-use std::io::{Read, Write};
+use std::io::{Cursor, Read, Seek, Write};
 
 /// The PNG magic header
 pub const PNG_HEADER: [u8; 8] = [137, 80, 78, 71, 13, 10, 26, 10];
+pub const IHDR_HEADER: [u8; 8] = [0, 0, 0, 13, 73, 72, 68, 82];
+const ASSUMED_ZTXT_MAX: usize = 500;
```
Comment (ZeWaka) [inline src/lib.rs:14]: why this number?

Comment (itsmeow) [reply src/lib.rs:14]: I looked at a typical DMI file and found that 500 was around the maximum number of bytes until the zTXt header ended. This number is only for performance reasons since it prompts the file reader for how many bytes to try to start by reading before it will need to read another chunk from the file. It's best for it to be smaller to reduce allocations, but large enough to not require most files to read more than once (so no smaller than the length of the file to the end of zTXt, on average)

File: tests/bench_dmi_load.rs (hunk 1 above)
Comment (itsmeow) [reply tests/bench_dmi_load.rs]: It's useful but I don't want it to run for regular CI because it uses external folders (ss13 icons folder). Not sure if I should just delete it then

File: src/lib.rs (hunk 2 above)
Comment (itsmeow) [reply src/lib.rs:14]: It ultimately has no impact on anything except performance because the reader will read additional bytes from the file if it tries to overread the in-memory buffer

**Review by ZeWaka — APPROVED** [review]

**Comment by LemonInTheDark** [comment]
unmentioned breaking change to image type grumble grumble
